### PR TITLE
Integrate protorpc client with GA library

### PR DESCRIPTION
**Pull request by shinfan**
Also creates a new example to demonstrate the usage.

### CODE REVIEW

File: client/client.go [hunk 1]
```diff
@@ -35,6 +36,25 @@ type Client struct {
...
+// will be used if no credentials are provided.
+// For more information about application default credentials please read:
+// https://cloud.google.com/docs/authentication/production
+func NewClient(ctx context.Context, credentials *sgauth.Credentials) (*Client, error) {
+	http, err := sgauth.NewHTTPClient(ctx, credentials)
+	if err != nil {
+		return nil, err
+	}
+	baseUrl := fmt.Sprintf("https://%s/$rpc/%s/",
+		credentials.ServiceAccount.ServiceName, credentials.ServiceAccount.APIName)
```
Comment (wora) [inline client/client.go]: This is not the abstraction that we should use. The authenticator should have no knowledge about the service or the API. We also should not provide convenient method until it is really needed. Here is what we should do.
```go
authenticator, err := sgauth.NewAuthenticator(ctx, nil)
if err != nil {
  return nil, err
}
return &Client{
  HTTP: authenticator.GetHTTPClient(),
  BaseURL: baseUrl, // from user input.
  UserAgent: "..."
}
```

File: example/googleauthenticator/main.go [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+	"context"
+	"github.com/shinfan/sgauth"
+
+	client "github.com/wora/protorpc/client"
+	proto "github.com/golang/protobuf/proto"
+	servicemanagement "google.golang.org/genproto/googleapis/api/servicemanagement/v1"
+)
+
+func NewClient(ctx context.Context, service_name string, api_name string) (*client.Client, error) {
+	var credentials = &sgauth.Credentials{
```
Comment (wora) [inline example/googleauthenticator/main.go]: I don't think we should call it Credentials. It should be a general concept of Settings.
```go
type Settings struct {
  // The JSON credentials downloaded from Google Cloud Console.
  // If unset, the library will use GOOGLE_APPLICATION_CREDENTIALS
  // environment variable to find the credentials in a file. If not found,
  // the library will fallback to the built-in service account provided by
  // the GCE/GKE/GAE runtime.
  Credentials string
  // Do we need this? Must be one of "OAuth2", "JWT".
  // Method string
  // OAuth 2 scope. According to the spec, this is a space-separate values.
  Scope string
  // JWT audience (aud). If not specified, the default one "https://{host}/{first-path-segment}".
  Audience string
  // Generates X-Goog-User-Project header.
  UserProject string
  // Generates X-Goog-Api-Key header.
  APIKey string
}
```

**Review by wora — COMMENTED** [review]
The current design has layering issue. The authenticator should only do authentication for requests, i.e., adding authentication related headers to the http request. It should have no knowledge about service or api or anything like that.

File: example/googleauthenticator/main.go (hunk 2 above)
Comment (shinfan) [reply example/googleauthenticator/main.go]: If we do not have `method` field, we cannot distinguish oauth and jwt. This can be ambiguous since "apikey" can be also considered as another method. Authentication methods have many dimensions and each dimension requires different parameters. This is why I prefer the hierarchic structure instead of a flattened one. 

Instead of passing a credential string, I suggest we make the name more specific to be `CredentialPath` so it is a file path.

`Audience` is a JWT specific concept and most users may not know how to construct. That's why I would suggest provide a helper to construct this string from host and api names for the users.

I suggest `Scope` to be a slice of strings, it is just easier to construct and more language idiomatic than a raw white-space separated string.

To conclude, I suggest something like this:
```go
type Settings struct {
  // OAuth2.0 method
  OAuth2 OAuth2
  // JWT method
  JWTToken JWT
  // Generates X-Goog-User-Project header.
  UserProject string
  // Generates X-Goog-Api-Key header.
  APIKey string
}

type OAuth2 struct {
  // The JSON credentials file downloaded from Google Cloud Console.
  // If unset, the library will use GOOGLE_APPLICATION_CREDENTIALS
  // environment variable to find the credentials in a file. If not found,
  // the library will fallback to the built-in service account provided by
  // the GCE/GKE/GAE runtime.
  CredentialsPath string
  // OAuth 2 scopes.
  Scope []string
}

type JWTToken struct {
  // The JSON credentials file downloaded from Google Cloud Console.
  // If unset, the library will use GOOGLE_APPLICATION_CREDENTIALS
  // environment variable to find the credentials in a file. If not found,
  // the library will fallback to the built-in service account provided by
  // the GCE/GKE/GAE runtime.
  CredentialsPath string
  // JWT audience(aud)
  // Or we can use a raw string and let user construct (?)
  Aud Audience
}

// Information used to construct the JWT audience.
type Audience struct {
    Host string
    APIName string
}
```

File: client/client.go (hunk 1 above)
Comment (julianshaoliu) [reply client/client.go]: "The authenticator should have no knowledge about the service or the API. " --- agree here. Although constructing audience offers convenience, it may leads to conceptual confusions.

Comment (shinfan) [reply client/client.go]: SGTM

File: example/googleauthenticator/main.go (hunk 2 above)
Comment (julianshaoliu) [reply example/googleauthenticator/main.go]: I am not convinced that the flattened structure is a good approach. If flatten it, it is hard to both explain to the user, and do sanity check internally, that Scope is needed for oauth2, audience is needed for jwt, and none needed for api key, etc. Also we will need a Method string to serve as the enum.

Why not use a embedded structure like below:

type OAuth2Token struct {
   CredentialsFile string
   Scope []string   /// or Scope string and choose concatenated format.
}

type JwtToken struct {
   CredentialsFile string
   Aud Audience // or String if we do not offer any construction convenience.
}

type ApiKey struct {
   ApiKey string
}

type Settings struct {
   Apikey ApiKey
   OAuth2Token  OAuth2Token
   JwtToken JwtToken
}

Then we do not need the Method. Whatever is not nil, indicates the method. A user only needs to pass one and leave others empty or nil. If a user passes multiple non-nil, we can either fail the request, or set a priority order.

Comment (wora) [reply example/googleauthenticator/main.go]: We probably should have a method field to explicitly select the auth method. I don't like the nested structure. It is just too complicated to use.

I strongly dislike the CredentialsFile. The application should read the file and pass json to the library, or we use GOOGLE_APPLICATION_CREDENTIALS environment variable.

Many people like to use string array for scope, I prefer to use a single string because that is what OAuth 2 spec uses. We should use whatever the spec says, regardless whether we like the spec or not.

For Google APIs, we almost never need to specify the audience. The library can easily calculate the audience based on the request url. We should remove it for now.
```go
type Settings struct {
  // We should leave this out until we need it. We only need to support
  // OAuth 2 and JWT for many years. The library can just guess itself.
  // Method string
  CredentialsJSON string
  // If specified, use OAuth. Otherwise, JWT.
  Scope string
  // This is only used for domain-wide delegation.
  User string
  // This name is confusing now. Since we have quotaUser and userProject.
  // We should have named them as quotaUser and quotaProject.
  QuotaUser string
  QuotaProject string
  APIKey string
}
```
NOTE: the user project and the api key unconditionally set the http headers. They are totally independent from others. The credentials is used if set, otherwise we do ADC. We could get rid of Method. If scope is set, we use oauth, otherwise we use jwt. JWT audience should be auto calculated for Google APIs. Users never need to specify it.

COMMENT: I don't like to deal with Set<String> or Vector<String> or []string for the scopes. We should force the application to join the string, so the library only deal with the single string. The same is for credentials. We force the app to read the file. For example, if the file is a URL, we don't know what permission or authentication is needed to fetch the file. or how to handle retry etc. So we better not to do it at all.

I believe the above struct should last for next 5 years without changing. We can add more fields later.

@bochunz

Comment (shinfan) [reply example/googleauthenticator/main.go]: Synced with @bochunz this morning and we reached an agreement that mostly aligns with your suggestion above:

1) Use flat structure instead of nested
2) Use JSON credential content string instead of file path
3) Whitespace-separated scope string instead of string slice
4) Use OAuth if scope is non-empty otherwise default to JWT.
5) Authenticator will check for required fields at runtime and returns error if anything missing.
6) Fallback to ADC if no settings is specified.

> For Google APIs, we almost never need to specify the audience. The library can easily
> calculate the audience based on the request url. We should remove it for now.

There are use cases that applications do not use a client. From layering perspective, the API client is on top of the auth lib. The auth lib has no knowledge about the url.  We probably still wanna provide the audience field so the value can be passed into the auth lib from upper layer. The API client can always calculate the value from the url, and set the value in this setting object for the users.

To summarize:

```go
type Settings struct {
  CredentialsJSON string
  // If specified, use OAuth. Otherwise, JWT.
  Scope string
  Audience string
  // Google API key
  APIKey string
  // This is only used for domain-wide delegation.
  User string
  // This name is confusing now. Since we have quotaUser and userProject.
  // We should have named them as quotaUser and quotaProject.
  QuotaUser string
  QuotaProject string
}
```

WDYT?

**Comment by shinfan** [comment]
I have updated the code based on the recent discussions. PTAL.

@wora @julianshaoliu

File: client/client.go [hunk 3]
```diff
@@ -35,6 +36,38 @@ type Client struct {
...
+func createClient(ctx context.Context, settings *sgauth.Settings, baseUrl string) (*Client, error) {
+	http, err := sgauth.NewHTTPClient(ctx, settings)
+	if err != nil {
+		return nil, err
+	}
+	return &Client{
+		HTTP:        http,
+		BaseURL:     baseUrl,
+		UserAgent:   "protorpc/0.1",
+		ApiKey: settings.APIKey,
```
Comment (wora) [inline client/client.go]: This should be removed. The auth client add API key automatically.

File: client/client.go [hunk 4]
```diff
@@ -35,6 +36,38 @@ type Client struct {
...
+	if settings == nil {
+		settings = &sgauth.Settings{}
+	}
+	if settings.AuthMethod() == sgauth.MethodJWT && settings.Audience == "" {
+		// Derive audience from url if using JWT and aud is not provided explicitly.
+		urlTokens := strings.Split(baseUrl, "/")
+		host := urlTokens[2]
+		api_name := urlTokens[4]
+		settings.Audience = fmt.Sprintf("https://%s/%s", host, api_name)
+		println(settings.Audience)
```
Comment (wora) [inline client/client.go]: We must auto generate the audience. The default format is https://{host}/{api}. If we set the audience, the authenticator cannot be used by another api, it forces application to create many authenticators, that really sucks.

File: client/client.go [hunk 5]
```diff
@@ -35,6 +36,38 @@ type Client struct {
...
+func NewClient(ctx context.Context, settings *sgauth.Settings, baseUrl string) (*Client, error) {
+	if settings == nil {
+		settings = &sgauth.Settings{}
+	}
+	if settings.AuthMethod() == sgauth.MethodJWT && settings.Audience == "" {
+		// Derive audience from url if using JWT and aud is not provided explicitly.
+		urlTokens := strings.Split(baseUrl, "/")
+		host := urlTokens[2]
+		api_name := urlTokens[4]
+		settings.Audience = fmt.Sprintf("https://%s/%s", host, api_name)
```
Comment (wora) [inline client/client.go]: Please check how grpc client handles this.

File: client/client.go (hunk 4 above)
Comment (shinfan) [reply client/client.go]: I think there are two possible approaches here of how client integrates with authenticator:

**OPTION 1)** 
Clients take authenticator as an parameter during construction. In this way user needs to create an authenticator first and then create API client. The code may look something like:

```go
settings := &sgauth.Settings{...}
authenticator := sgauth.NewAuthenticator(settings)
client := client.NewClient(authenticator, baseUrl)
``` 
Pros:
-  The authenticator instance can be reused

Cons:
- More complex code comparing with option 2)

**OPTION 2)** 
Clients take the Settings object directly and create authenticator for the user.
```go
settings := &sgauth.Settings{...}
client := client.NewClient(settings, baseUrl)
``` 

Pros:
- This code is simpler than option 1)

Cons:
- Every client has its authenticator instance, it is not shared. But since users do not need to construct authenticator themselves, this is not a user burden --- just a little extra memory usage.

NOTE: With __sharded google__, users may need multiple authenticators anyway to support multi-region since authenticator itself is shard-agnostic.

**Comment by wora** [comment]
For 3LO, you don't want to login N times. It will drive you nuts. We must
share authenticator.

**Comment by wora** [comment]
Technically, the authenticator is an advanced HTTP client that knows how to
inject credentials into HTTP requests and gRPC requests. The credentials
include most system parameters supported by Google APIs.

File: client/client.go [hunk 6]
```diff
@@ -35,6 +36,34 @@ type Client struct {
 
 }
 
+// Creates a client with the given credentials. Application default credentials
+// will be used if no credentials are provided.
+// For more information about application default credentials please read:
+// https://cloud.google.com/docs/authentication/production
+func NewClient(ctx context.Context, settings *sgauth.Settings, baseUrl string) (*Client, error) {
```
Comment (wora) [inline client/client.go]: This design binds the client library with the auth library. That is not a great design choice. We should force application to create authenticator first, then create client, so the client library works with other authenticators.

authenticator := NewAuthenticator(settings)
client := NewClient(authenticator.NewHTTPClient, baseUrl)

File: client/utils.go [hunk 7]
```diff
@@ -0,0 +1,14 @@
+package client
+
+import (
+	"strings"
+	"fmt"
+)
+
+// Derive the audience from base URL
+func createAudience(baseUrl string) (string) {
+	urlTokens := strings.Split(baseUrl, "/")
```
Comment (wora) [inline client/utils.go]: Let's remove this logic for now and force application to specify the audience.

Comment (shinfan) [reply client/utils.go]: SGTM.

GRPC-GO uses similar logic to create the audience from the host and method:
https://github.com/grpc/grpc-go/blob/339b6cb107199486c3b352f70db9e977191be8b7/internal/transport/http2_client.go#L455

File: client/client.go (hunk 6 above)
Comment (shinfan) [reply client/client.go]: Decoupling the client and authenticator sounds reasonable.
I was thinking of minimize the initialization code :)

Comment (shinfan) [reply client/client.go]: I have simplified this to just use:
```go
        http, err := sgauth.NewHTTPClient(context.Background(), &sgauth.Settings{
		Audience: os.Args[1],
	})
	if err != nil {
		fmt.Println(err.Error())
	}
	c:= &client.Client{
		HTTP:        http,
		BaseURL:     os.Args[2],
		UserAgent:   "protorpc/0.1",
	}
```
so we do not need to change anything in current client.go

**Comment by shinfan** [comment]
Comments addressed. PTAL.

FYI I implemented two example code. One of them is in this PR.
The other one is in the Google Authenticator repo:
https://github.com/shinfan/sgauth/tree/master/example/library

**Review by wora — APPROVED** [review]
